h/views/api/group_members.py

```diff
@@ -90,4 +90,12 @@ def edit_member(context: GroupMembershipContext, request):
             old_roles,
         )
 
-    return GroupMembershipJSONPresenter(context.membership).asdict()
+    if context.user == request.user:
+        # Update request.identity.user.memberships so permissions checks done
+        # by GroupMembershipJSONPresenter below return the right results.
+        # Otherwise permissions checks will be based on the old roles.
+        for membership in request.identity.user.memberships:
+            if membership.group.id == context.group.id:
+                membership.roles = new_roles
```

Comment:
The problem here is that when a user edits their own membership that edit-membership API's response body contains the updated version of the user's own membership in the response, and obviously we want that to be up-to-date with the newly-changed roles and actions. But if we just go ahead and let `GroupMembershipJSONPresenter` use `request.has_permission()` to query what those new roles and actions in the response should be it's going to query based on the user's _old_ memberships and roles because `request.identity` is a `LongLivedUser` object whose `memberhips` is a list of `LongLivedMembership` objects that're created once the first time that `Identity.from_models()` is called in the request.

So here I'm mutating `request.identity.user.memberships` to make the membership dict in the response come out right.

This may be the least invasive workaround.

An alternative would be to have `GroupJSONPresenter` create an updated `Identity` object and call `identity_permits()` directly instead of calling `request.has_permission()`. So bypassing the outdated `request.identity` instead of mutating it. But creating a new `Identity` with `Identity.from_models(user)` will fetch all of the user's memberships from the DB. We might get away with that. If we wanted to avoid it we'd need to construct a `LongLivedUser` with a list of `LongLivedMembership`'s that's just the user's membership in `context.group`. Something like this should work:

```python
identity = Identity(
    user=LongLivedUser(
        id=self.membership.user.id,
        userid=self.membership.user.userid,
        authority=self.membership.user.authority,
        staff=self.membership.user.staff,
        admin=self.membership.user.admin,
        memberships=[],
    )
)
identity.memberships.append(
    LongLivedMembership(
        group=LongLivedGroup(
            id=self.membership.group.id,
            pubid=self.membership.group.pubid,
        ),
        user=identity.user,
        roles=self.membership.roles,
    )
)

if identity_permits(identity, context, permission):
    ...
```

This would duplicate quite a lot of code from `identity.py` for constructing `LongLivedMembership`'s, `LongLivedGroup`'s, `LongLivedUser`'s and `Identity`'s from models so we might want to refactor `Identity.from_models()` to accept the list of memberships as an argument instead of reading them from the DB itself, enabling something like:

```python
identity = Identity.from_models(user, memberships=[self.membership])`
```